segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ConcatOperation.java

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.Exceptions;
+import io.pravega.common.LoggerHelpers;
+import io.pravega.common.Timer;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import io.pravega.segmentstore.contracts.StreamSegmentTruncatedException;
+import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.StorageNotPrimaryException;
+import io.pravega.segmentstore.storage.metadata.ChunkMetadata;
+import io.pravega.segmentstore.storage.metadata.MetadataTransaction;
+import io.pravega.segmentstore.storage.metadata.SegmentMetadata;
+import io.pravega.segmentstore.storage.metadata.StorageMetadataWritesFencedOutException;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+
+import static io.pravega.segmentstore.storage.chunklayer.ChunkStorageMetrics.SLTS_CONCAT_COUNT;
+import static io.pravega.segmentstore.storage.chunklayer.ChunkStorageMetrics.SLTS_CONCAT_LATENCY;
+
+/**
+ * Implements the concat operation.
+ */
+@Slf4j
+class ConcatOperation implements Callable<CompletableFuture<Void>> {
+    private final long traceId;
+    private final SegmentHandle targetHandle;
+    private final long offset;
+    private final String sourceSegment;
+    private final ChunkedSegmentStorage chunkedSegmentStorage;
+    private final ArrayList<String> chunksToDelete = new ArrayList<>();
+
+    private volatile Timer timer;
+    private volatile String targetSegmentName;
+    private volatile SegmentMetadata targetSegmentMetadata;
```

Comment:
This should not matter as much, I used operation class so that I don't have to pass lots of parameters around. passing parameters becomes troublesome specially when multiple values need to be returned. These field are modified and then used in next step.